core/Enclave/key_operation.cpp

```diff
@@ -0,0 +1,2044 @@
+    {
+        if (EVP_DecryptFinal_ex(pState, plaintext->data + temp_len, &temp_len) <= 0)
+        {
+            printf("Error: fail to finalize the decryption\n");
+            ret = SGX_ERROR_UNEXPECTED;
+            goto out;
+        }
+    }
+out:
+    if (pState != NULL)
+    {
+        EVP_CIPHER_CTX_free(pState);
+    }
+    memset_s(dec_key, sizeof(dec_key), 0, sizeof(dec_key));
+    SAFE_FREE(dec_key);
+    return ret;
+}
+
+sgx_status_t ehsm_rsa_encrypt(const ehsm_keyblob_t *cmk,
+                              const ehsm_data_t *plaintext,
+                              ehsm_data_t *ciphertext)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    //verify padding mode
+    if (cmk->metadata.padding_mode != EH_PAD_RSA_PKCS1 && cmk->metadata.padding_mode != EH_PAD_RSA_PKCS1_OAEP)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint8_t *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *rsa_pubkey = NULL;
+
+    // make sure this key pair can work
+    // ret = ehsm_judge_rsa_keypair_available(cmk);
+    // if (ret != SGX_SUCCESS)
+    // {
+    //     goto out;
+    // }
+
+    // load rsa public key
+    rsa_keypair = (uint8_t *)malloc(cmk->keybloblen);
+    if (rsa_keypair == NULL)
+    {
+        goto out;
+    }
+
+    ret = ehsm_parse_keyblob(rsa_keypair, cmk->keybloblen,
+                             (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    if (ret != SGX_SUCCESS)
+        goto out;
+
+    bio = BIO_new_mem_buf(rsa_keypair, -1); // use -1 to auto compute length
+    if (bio == NULL)
+    {
+        printf("failed to load public key pem\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    // make encryption
+    PEM_read_bio_RSA_PUBKEY(bio, &rsa_pubkey, NULL, NULL);
+    if (rsa_pubkey == NULL)
+    {
+        printf("failed to load rsa key\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    if (ciphertext->datalen == 0)
+    {
+        ciphertext->datalen = RSA_size(rsa_pubkey);
+        ret = SGX_SUCCESS;
+        goto out;
+    }
+    if (RSA_public_encrypt(plaintext->datalen, plaintext->data, ciphertext->data, rsa_pubkey, cmk->metadata.padding_mode) != RSA_size(rsa_pubkey))
+    {
+        printf("failed to make rsa encryption\n");
+        goto out;
+    }
+
+    ret = SGX_SUCCESS;
+out:
+    BIO_free(bio);
+    RSA_free(rsa_pubkey);
+
+    memset_s(rsa_keypair, cmk->keybloblen, 0, cmk->keybloblen);
+    SAFE_FREE(rsa_keypair);
+
+    return ret;
+}
+
+sgx_status_t ehsm_sm2_encrypt(const ehsm_keyblob_t *cmk,
+                              const ehsm_data_t *plaintext,
+                              ehsm_data_t *ciphertext)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    uint8_t *sm2_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *sm2_pubkey = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ectx = NULL;
+
+    // load sm2 public key
+    sm2_keypair = (uint8_t *)malloc(cmk->keybloblen);
+    if (sm2_keypair == NULL)
+    {
+        goto out;
```

Comment:
Please confirm below functions can be well handled if all are NULL.
    BIO_free(NULL);
    RSA_free(NULL);
    EVP_PKEY_free(NULL);
    EVP_PKEY_CTX_free(NULL);

I juest checked BIO_free is OK to handle NULL.  "If a is NULL nothing is done." https://www.openssl.org/docs/man3.0/man3/BIO_free.html

Reply:
otherwise, you have to check if NULL before free at out.

Reply:
please check all other similar issues.

Reply:
it will be changed